### ISSUE TITLE
Python:  Create VImage from string/buffer?

### ISSUE THREAD

**Issue by stepchowfun** [libvips/libvips#23]
From the documentation, I see these two functions:

```
VImage.fromstring (string,  
  width, height, bands, format) ->  
  VImage
VImage.frombuffer (buffer,  
  width, height, bands, format) ->  
  VImage
```

Can I use these functions to load, say, a JPEG which is stored in memory as a string/buffer?  If so, the issue is that I don't know the size/bands/format beforehand.  Basically what I want is something like the usual constructor which takes a filename, except I want to pass in a buffer instead of a filename (or if I could pass in a file descriptor/pipe, that would work for me as well).  Is this not available in the Python binding?  Thanks!

**Comment by scyclops** [comment]
Any update on if/when the Python bindings will support jpegload_buffer (as well as gifload_buffer and pngload_buffer)?

**Comment by jcupitt** [comment]
Hi @scyclops, yes, the new Python binding was released with version 8 in 2015. 

http://jcupitt.github.io/libvips/API/current/using-from-python.html

Use it like this:

```
$ python
Python 2.7.13 (default, Jan 19 2017, 14:48:08) 
[GCC 6.3.0 20170118] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import gi
>>> gi.require_version('Vips', '8.0')
>>> from gi.repository import Vips
>>> with open("pics/k2.jpg", 'rb') as f:
...     str = f.read()
... 
>>> image = Vips.Image.new_from_buffer(str, "")
>>> image.avg()
102.80430080369972
```

The extra `""` argument is for any load options you might want to set as strings. 

There it's sniffing the file type from the first few bytes of the string and calling the appropriate loader for you. You can also call the jpeg loader directly:

```
>>> image = Vips.Image.jpegload_buffer(str)
>>> image.width
1450
>>> image.avg()
102.80430080369972
```

You can give optional arguments, for example:

```
>>> image = Vips.Image.jpegload_buffer(str, shrink = 2, access = Vips.Access.SEQUENTIAL)
>>> image.width
725
>>> image.avg()
102.80430080369972
```

Now it's loading in sequential mode, and at half size. You can give extra args to `.new_from_buffer()` as well.

There are quite a few more buffer loaders now:

```
$ vips -l | grep load_buffer
          VipsForeignLoadPdfBuffer (pdfload_buffer), load PDF with libpoppler, priority=0, is_a_buffer, get_flags, get_flags_filename, header, load
          VipsForeignLoadSvgBuffer (svgload_buffer), load SVG with rsvg, priority=0, is_a_buffer, get_flags, get_flags_filename, header, load
          VipsForeignLoadGifBuffer (gifload_buffer), load GIF with giflib, priority=0, is_a_buffer, get_flags, get_flags_filename, header
        VipsForeignLoadPngBuffer (pngload_buffer), load png from buffer, priority=0, is_a_buffer, get_flags, header, load
          VipsForeignLoadJpegBuffer (jpegload_buffer), load jpeg from buffer, priority=0, is_a_buffer, get_flags, header, load
          VipsForeignLoadWebpBuffer (webpload_buffer), load webp from buffer, priority=-50, is_a_buffer, get_flags, header, load
          VipsForeignLoadTiffBuffer (tiffload_buffer), load tiff from buffer, priority=0, is_a_buffer, get_flags, header, load
          VipsForeignLoadMagickBuffer (magickload_buffer), load buffer with ImageMagick, priority=-100, is_a_buffer, get_flags, get_flags_filename, header
```

**Comment by scyclops** [comment]
Great - it looks like I may have been using the old python binding that I installed by running the below on Ubuntu 16.04.

`apt-get install python-vipscc`

Any clue on how to easily install the new python bindings on Ubuntu 16.04?  It looks like running the following on Ubuntu 16.04 installs vips 8.2.2 with the old bindings but the typelib file is nowhere to be found after the install and *from gi.repository import Vips* always fails.  Or is compiling from source the the only way to get the typelib installed?

`apt-get install libvips libvips-tools python-vipscc gobject-introspection python-gi --no-install-recommends`

**Comment by jcupitt** [comment]
Hi, `python-vipscc` is the old binding, you don't need to install it, unless you really want to use it.

The typelib is in `gir1.2-vips-8.0`. It used to not include the overrides file, unfortunately. I filed a bug, but I don't know if it was fixed for 16.04.

If you are missing the Vips.py overrides, you can download here:

https://github.com/jcupitt/libvips/blob/master/python/packages/gi/overrides/Vips.py

There's a note in the comment at the top saying where to copy the file to.
